File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    var answer = 0;
    let left = 1, right = 0;
    
    for(diff of diffs) right = Math.max(right,diff);
    answer = right;
    
    while(left<=right){
        const level = parseInt((left+right)/2);
        let sum = times[0];
        
        for(let i = 1; i < diffs.length;i++){
            if(level>=diffs[i]) sum+=times[i];
            else sum += (times[i]+times[i-1])*(diffs[i]-level)+times[i];
        }
        
        if(sum <= limit){
            answer = Math.min(answer,level);
            right = level-1;
        }else left = level+1;
    }
    return answer;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        int answer = 0;
        int right = 0, left = 1;
        int n = diffs.length;
        
        for(int diff : diffs) right = Math.max(right,diff);
        answer = right;
        
        while(left<=right){
            int level = (left+right)/2;
            long sum = times[0];
            
            for(int i=1;i<n;i++){
                if(level>=diffs[i]) sum+=(long)times[i];
                else sum+= ((times[i]+times[i-1])*(diffs[i]-level)+times[i]);
            }
            
            
            if(sum <= limit){
                answer = Math.min(answer, level);
                right = level-1;
            }
            else left = level+1;
            
        }
        return answer;
    }
}